### ISSUE TITLE
Set value

### ISSUE THREAD

**Issue by Duxez** [andrerpena/react-mde#87]
At a previous version I could just set the object in state, but that doesn't seem to work after updating to the latest release.

I had to update since the editor forcefully took the focus. Does anyone have an example on how to set a default value for the editor?

**Comment by andrerpena** [comment]
Oh. This is actually very easy but you reminded me that I now have to provide an utility for it... But I can do it in 20 minutes.

I'll update the project this night, as well as the documentation and post an example here.

Thank you for informing me.

**Comment by andrerpena** [comment]
Please take a look at this example where a button is used to duplicate the text: https://codesandbox.io/s/l4jlp3p5om

If you need any further help, please re-open.

**Comment by andrerpena** [comment]
In this case it's fine. If you pass null as a `editorState`, `React-mde` will automatically create an `MdeState` and call your `onChange` handler with a brand new object for you. You can then use `buildNewMdeState` from this point on.

It's important to state that the reason why you neeed a `buildNewMdeState` instead of simply setting some text, is that 1) You need to generate the HTML for your new markdown. 2) The new state must be included in the history so Ctrl/Cmd + Z will work. Unfortunately stuff are not as simple as we wanted them to be.

**Comment by andrerpena** [comment]
I'm updating the docs to refer to this thread to help future people

**Comment by Duxez** [comment]
Thanks for your help that did the trick!

**Comment by smoleniuch** [comment]
@andrerpena You can not pass `editorState` as null.[This](https://github.com/andrerpena/react-mde/blob/master/src/util/DraftUtil.ts#L134) throws an error when i tried to do this.You must pass `{html:'', markdown:'', draftEditorState: null }` as default.

Cheers!

**Comment by andrerpena** [comment]
@smoleniuch I've created an issue for this and I'll take a look this night. Thanks for reporting and for providing a Sanbox.

**Comment by andrerpena** [comment]
I'm sorry for how long it took for me to fix this. I recently moved to Berlin and I haven't had much to code at home. But now things are calming down and I'm returning.

There was a bug that prevented you from changing the `editorState.markdown` property. Now it is fixed.
If you want to create a button that duplicates text, just create it and add this as the event handler:

```
handleButtonClick = () => {
        this.setState({
            mdeState: {
                markdown: this.state.mdeState.markdown + "\n" + this.state.mdeState.markdown,
            },
        });
    }
```

**Comment by orditeck** [comment]
I feel like I'm having a similar issue.

I'm trying to do the following:

1. Render ReactMde with an empty state `{ draftEditorState: null, html: '', markdown: '' }`
2. Fetch the content remotely
3. When done, set the content in the store and refresh ReactMde

Expected result: empty ReactMde while loading, then ReactMde filled with the content after loading.

Result: empty ReactMde. It doesn't seem to refresh/re-render when the state changes. If I unmount/remount the component, the content is there in the first example below. So it's really when refreshing the state from outside the component that it doesn't work. Any idea why? I'm probably doing something wrong here as I saw no issue about this.

Both

```js
export default EditorStore.subscribe(
    class extends Component {
        state = {
            mdeState: {
                markdown: this.props.todo.content
            }
        };

        handleValueChange = mdeState => this.setState({ mdeState: mdeState });

        render() {
            return (
                <ReactMde
                    onChange={this.handleValueChange}
                    editorState={this.state.mdeState}
                    generateMarkdownPreview={markdown =>
                        Promise.resolve(MarkdownConverter.makeHtml(markdown))
                    }
                />
            );
        }
    }
);
```

and 

```js
export default EditorStore.subscribe(
    class extends Component {
        handleValueChange = mdeState =>
            EditorStore.set({
                mdeState: mdeState
            });

        render() {
            return (
                <ReactMde
                    onChange={this.handleValueChange}
                    editorState={this.props.mdeState}
                    generateMarkdownPreview={markdown =>
                        Promise.resolve(MarkdownConverter.makeHtml(markdown))
                    }
                />
            );
        }
    }
);
```

doesn't work.

This is how it get refreshed from another component:

```js
Todos.find(this.props.match.params.id).then(({ data }) =>
    EditorStore.set({
        loading: false,
        todo: data.data,
        mdeState: {
            ...EditorStore.get('mdeState'),
            markdown: data.data.content
        }
    })
);
```

**Comment by orditeck** [comment]
I just tried with the master branch instead of the latest npm release, same issue. Any idea what I'm doing wrong @andrerpena 😕?

**Comment by andrerpena** [comment]
@orditeck sorry!!! I'm on it now!

**Comment by andrerpena** [comment]
@orditeck I tried reproducing the problem and I could not... I'll give you some hints and then I'll post a working code that updates the content asynchronously, on a `setTimeout`.

**hints**:

The `editorState` that you pass to React-Mde is considered to be initialized when it's internal `draftEditorState` is truthy. If you pass a falsy `draftEditorState`, it will initialize it internally and call the `onChange` callback with the initialized value. 

The only thing I can think of, that you could be doing wrong, is to pass a non falsy internal `draftEditorState`.

This is a working example:

```js
import * as React from "react";
import ReactMde, { ReactMdeTypes } from "../src";
import * as Showdown from "showdown";

export interface AppState {
    mdeState: ReactMdeTypes.MdeState;
}

export class App extends React.Component<{}, AppState> {

    converter: Showdown.Converter;

    constructor(props) {
        super(props);
        this.state = {
            mdeState: {
                markdown: "",
            },
        };
        this.converter = new Showdown.Converter({
            tables: true,
            simplifiedAutoLink: true,
            strikethrough: true,
            tasklists: true,
        });
        setInterval(() => {
            this.setState({
                mdeState: {
                    markdown: this.state.mdeState.markdown + " Hello World!",
                },
            });
        }, 2000);
    }

    handleValueChange = (mdeState: ReactMdeTypes.MdeState) => {
        this.setState({ mdeState });
    }

    render() {
        return (
            <div className="container">
                <ReactMde
                    onChange={this.handleValueChange}
                    editorState={this.state.mdeState}
                    generateMarkdownPreview={(markdown) => Promise.resolve(this.converter.makeHtml(markdown))}
                />
            </div>
        );
    }
}

```

**Comment by orditeck** [comment]
Thanks @andrerpena for your help.

It was working with my workaround and it wasn't bad, but still I wanted to try the "right" way. So I reverted back to the version on npm and retried to make it work using your code as an example. It ended up working just like I expected the first time, 2 days ago. No idea why...

Final code looks like

```js
export default EditorStore.subscribe(
    class extends Component {
        render() {
            return (
                <ReactMde
                    onChange={mdeState =>
                        EditorStore.set({
                            mdeState: mdeState
                        })
                    }
                    editorState={this.props.mdeState}
                    generateMarkdownPreview={markdown =>
                        Promise.resolve(MarkdownConverter.makeHtml(markdown))
                    }
                />
            );
        }
    }
);
```

The default `EditorStore.mdeState` is simply

```js
mdeState: {
    markdown: ''
}
```

And when I fetch the content, I set the state to:

```js
Todos.find(this.props.match.params.id).then(({ data }) =>
    EditorStore.set({
        mdeState: {
            markdown: data.data.content
        }
    })
);
```

And it just works. No idea what I did wrong two days ago.

**Comment by bnbarak** [comment]
Thank you for @andrerpena for taking the time to solve those issues. 
## Issue
`<ReactMde />` not updating when there is change in state.

### Reproduce code
```
export default class MarkdownEditor extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      mdeState: {
        markdown: props.markdown || ""
      }
    };

    this.handleValueChange = this.handleValueChange.bind(this)
  }

  static getDerivedStateFromProps(props, state) {
    const newState = state;
    newState.mdeState = {
      ...state.mdeState,
      markdown: props.markdown
    };
    return newState;
  }

  handleValueChange = (mdeState) => {
    this.setState({mdeState});
    this.props.handleMarkdownChange(mdeState);
  };

  render() {
    console.log(this.state.mdeState) // <-- prints the updated data
    return <ReactMde
        onChange={this.handleValueChange}
        editorState={this.state.mdeState}
        commands={commands}
        buttonContentOptions={{
          iconProvider: name => <Icon icon={name}/>,
        }}
      />
  }
}
```

### Note
With adding this (random) interval it works:
```
setInterval(() => {
      this.setState({
        mdeState: {
          markdown: this.state.mdeState.markdown,
        },
      });
    }, 2000);
```

**Comment by ryanlward** [comment]
I've also been experiencing this issue, using 5.7.0

One odd behavior I have seen is that is I set state using a string, it seems to display, but if I pull it form props it does not.

Ex: Doesn't work
```
constructor(props) {
        super(props);
        const value = ;
        this.state = {
            mdeState: { markdown: props.markdown || '' }
        };
        ...
}
```

Ex: Works
```
constructor(props) {
        super(props);
        const value = ;
        this.state = {
            mdeState: { markdown: "Foo" }
        };
        ...
}
```

Not sure why that would be any different.

**Comment by orszaczky** [comment]
In which version has this been resolved? I'm running 5.7.0 and have the same occurrence as above (setting the state as a string works, and the editor picks up the state, but if I try to pass it in from a prop, it doesn't.

I also tried updating the state with setInterval() which updates the component state, but the editor is not picking up the change.
